nupic/research/frameworks/pytorch/hooks/hook_manager.py

```diff
@@ -0,0 +1,138 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2020, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from .base import TrackStatsHookBase
+
+
+class ModelHookManager:
+    """
+    This class registers and manages a set of hooks of subclassed from
+    `TrackStatsHookBase`. The given hook is registered on all modules within
+    'named_modules'.
+
+    Tracking is started and stopped for all hooks via `self.start_tracking()` and
+    `self.stop_tracking()`. Alternatively, this class can be used a context manager to
+    automate these calls. For example,
+
+    ```
+    with hook_manager as hooks:
+        ... # Train here
+         stats = hooks.get_statitics()
+    ```
+
+    .. _filter_modules: nupic.research.frameworks.pytorch.model_utils.filter_modules
+
+    :param named_modules: dict mapping names to modules
+    :param hook_class: class subclassed from `TrackStatsHookBase`
+    :param hook_type: whether to register the hook as "forward" or "backward"
+                      or "pre_forward"
```

Comment:
the parameter specification should be in the docstring for the `__init__` method. In fact, most classes you define have the initialization parameters defined in the class docstring. Please make the appropriate changes throughout all files.

Reply:
That is not my interpretation of PEP 257. What we might have is some inconsistency throughout our repo. 

> The docstring for a class should summarize its behavior and list the public methods and instance variables. If the class is intended to be subclassed, and has an additional interface for subclasses, this interface should be listed separately (in the docstring). The class constructor should be documented in the docstring for its __init__ method. Individual methods should be documented by their own docstring.

Reply:
From what I could gather Sphinx follow that convention. Document public attributes in class docstrings, and the constructor in the init function (google also follow that same convention). So there might be some redundancy. 

Here are some examples: https://www.sphinx-doc.org/en/master/usage/extensions/example_google.html

Reply:
I see @mvacaporale is following the same convention as Sphinx, and according to the blame history for [sparse weights](https://github.com/numenta/nupic.torch/blob/master/nupic/torch/modules/sparse_weights.py), @lscheinkman also followed the same convention. Looks like this is indeed the way to go.